# References & Sources
Here we list all liturature and software with a brief summarizations.

<hr>

[Identification and Characterization of Multi-Species
Conserved Sequences](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC403793/pdf/0132507.pdf) <br>
<i>Elliott H. Margulies et al. 2003</i><br>
From this paper we derive ideas for tools relating to conserved sequence computation.
- [footprinter](https://www.ncbi.nlm.nih.gov/pubmed/12824433?dopt=Abstract): A program designed for phylogenetic footprinting.


[Phylogenetic Analysis by Maximum Likelihood (PAML)](http://abacus.gene.ucl.ac.uk/software/paml.html#PAMLx)<br>
bioinformatics tool for phylogenetic analysis

[Citing the Ensembl Database](http://www.ensembl.org/info/about/publications.html)<br>
information provided by ensembl on how to cite their database.

[Distinguishing-Homologous-from-Analogous-Proteins](https://academic.oup.com/sysbio/article-abstract/19/2/99/1655771/Distinguishing-Homologous-from-Analogous-Proteins?redirectedFrom=fulltext)<br>
<i>Walter M. Fitch, 1970</i><br>
A core paper on the topic of homology and orthologs being one of the first published on this discussion.

[MUSCLE Alignment software](http://www.drive5.com/muscle/)<br>
<i>Edgar R.C., 2004</i><br>
This tool is our primary resource for computing multiple sequence alignments. info for [citing MUSCLE](http://www.drive5.com/muscle/manual/papers.html)

[NCBI blast methods and algorithm overview](https://www.ncbi.nlm.nih.gov/books/NBK1734/)<br>
<i>David Wheeler and Medha Bhagwat, 2007</i><br>
In Comparative Genomics volume 1 and 2, statistical significance of blast results is discussed. Values of 0.001 to 0.0000001 are commonly used according to this resource.
